File: 小码哥大神5期练习/5、Runtime&RAC&MVVM&新特性/1025/函数编程思想的计算器/函数编程思想的计算器/Calculation.m
//
//  Calculation.m
//  函数编程思想的计算器
//
//  Created by coco on 16/7/18.
//  Copyright © 2016年 XHJ. All rights reserved.
//

#import "Calculation.h"

@implementation Calculation

- (instancetype)Calculation:(float (^)(float))block {
    if (block) {
        _result = block(_result);
    }
    return self;
}

@end

File: 小码哥大神5期练习/5、Runtime&RAC&MVVM&新特性/1025/01-Block/01-Block/Calculation.swift
//
//  Calculation.swift
//  01-Block
//
//  Created by M-coppco on 16/7/17.
//  Copyright © 2016年 M-coppco. All rights reserved.
//

import UIKit

class Calculation: NSObject {
    var result: Float = 0
    func a(number: Float) ->Calculation {
        self.result += number
       return self
    }
    
    func b(number: Float) ->Calculation {
        self.result -= number
        return self
    }
    
    func c(number: Float) ->Calculation {
        self.result *= number
        return self
    }
    
    func d(number: Float) ->Calculation {
        self.result /= number
        return self
    }
}


extension Calculation {
    class func makeCalculation(closure:(Calculation) -> Void) -> Float {
        let make: Calculation = Calculation()
        closure(make)
        return make.result
    }
}
